File: lib/node/core/src/Middleware.ts
import { Message } from "Message";
import { Provider } from "./Provider";

export type MiddlewareHandler = (state: object, message: Message) => object;

export abstract class Middleware {
    protected provider?: Provider;

    /**
     * Called on the store initialized.
     */
    public onInitialized(provider: Provider): void {
        this.provider = provider;
    }


    public handle(
        state: object,
        message: Message,
        next: MiddlewareHandler
    ): object | null {
        return next(state, message);
    }

    public dispose() {

    }
}

File: src/Memento.Core/Middleware.cs
namespace Memento.Core;

public abstract class Middleware : IDisposable {
    MiddlewareHandler? _handler;

    /// <summary>
    /// Gets the handler for this middleware.
    /// </summary>
    public MiddlewareHandler Handler => _handler
        ?? throw new InvalidOperationException($"Middleware '{GetType().FullName}' has not initialized.");

    /// <summary>
    /// Initializes the middleware asynchronously.
    /// </summary>
    /// <param name="provider">The service provider.</param>
    internal async Task InitializeAsync(IServiceProvider provider) {
        var handler = Create(provider);
        _handler = handler;
        await handler.InitializedAsync();
    }

    /// <summary>
    /// Creates the handler for this middleware.
    /// </summary>
    /// <param name="provider">The service provider.</param>
    /// <returns>The created middleware handler.</returns>
    protected abstract MiddlewareHandler Create(IServiceProvider provider);

    /// <summary>
    /// Disposes the middleware.
    /// </summary>
    public void Dispose() {
        _handler?.Dispose();
    }
}
